content/posts/numpy/generator-rng/index.md

```diff
@@ -0,0 +1,304 @@
+sequence of points, there are other technics that would generate almost
+perfect histograms only using a fraction of the number of points which is used
+here. See [this article]({{< relref "../../scipy/qmc-basics" >}})
+for more information on this.
+
+But why does it matter? How bad this is will depend on your application.
+If you don't, statistical properties are lost. This is hard to get because for
+most people, random is random. So taking an arbitrary draw of sample should not
+matter. And it does not **if** you replicate your analysis using different
+**random** seeds. This is the part everyone is forgetting. Whether you are
+optimizing something or calculating some statistics, you could get (un)lucky
+if you do this just once.
+
+If your application needs random numbers, then do use random numbers and play
+with the fact that they are random. You do science or anything critical
+using these randomly generated numbers, then please stay away from seeds.
+
+Some will argue that seeding is just used by developers for testing, well it is
+not. People don't like uncertainties. They naturally try to "fix" this
+"issue" by using a seed in their systems, and not just for testing.
+The worst part here is that people tend to use the same numbers. Some
+copy/pasting might be at play there as well.
+
+Think about it for a while. If everybody is using a seed, which is the same,
+and the original intent was to interact with random numbers for statistical
+analysis or optimization. Do you see the bias that is being introduced?
+
+## Testing with random numbers
+
+Before showing any code though, ask yourself if you actually need ot use
+random numbers to write your test. You might just need a small matrix with
+hard-coded numbers. Not only you don't need to run the code to know the values,
+but you will be 100% certain that this is stable on any platform/architecture
+and future-proof. You have more advanced needs, ok then maybe using known
+sequences (quasi-random numbers) is a good alternative (again see
+[here]({{< relref "../../scipy/qmc-basics" >}})
+).
+
+If you want or need to use NumPy for that, then either use
+`np.random.Generator` or the legacy `np.random.RandomState`. Aside from
+backward compatibility concerns, `np.random.RandomState` is still present in
+NumPy for testing.
+
+```python
+rng = np.random.default_rng()
+# or
+rng = np.random.RandomState()
+```
+
+[NEP19](https://numpy.org/neps/nep-0019-rng-policy.html) actually
+recommends `np.random.RandomState` for testing as the byte stream is not
+guaranteed as strongly as before with the new generators. In practice, changes
+are not happening often though. What matters in both cases is that you use the
+object based API and not the legacy approach with global state.
+
+Now, one thing with tests is that we want them to be reproducible so that
+we can inspect our code and find issues. Here you want to use a seed.
+This is the **only** valid use for a seed. (Sure if you are an expert in
```

Comment:
I don't agree with this. For example, in academic poblishing you might need to provide code the reproduces statistical methods that rely on Monte Carlo methods (e.g., the Bootstrap). In these cases you also need seed.  The key here IMO is to use a high-quality as-if random 256bit seed, which is like calling `default_rng()` and then storing the value used to seed the first `Generator` and then reusing this when reproducing.

Reply:
The last example I can think of that I had, was using a t-SNE (a stochastic embedding) on a smaller dataset.  Of course I have to make sure that the analysis is robust, but I still want to be able to plot an example and be able to replot it without getting a totally different image.

And for larger numerical studies I had intermediate results so large that I did not keep them.  But I still want to be able to "go back" and look at those intermediate results (effectively videos) in case there was something interesting/odd.  So seeding is necessary (although it can be a truly random seed!).
And even months of computer time used, doesn't mean that the final result won't have just few data-points.  So quantitative reproduction would still require seeding (and using the same software version etc.)...

I wonder what the "middle ground" is, since I suspect there is a good reason for the strong opinion:
* You can probably recommend high quality (even truly random) seeds in general.  Which will also prevent cargo-culting a seed into every script, since it requires first generating a seed (i.e. a two step process).
* It may make sense to recommend avoiding to seed until a need for reproducibility is clear?  That way, if you re-run an analysis anyway, you will naturally learn how much the results depends on the random stream.  Arguable, it is mainly a trick to avoid forgetting that even the random stream may be a source of bias(?) that needs to be considered.